modnet/preprocessing.py

```diff
@@ -274,277 +447,413 @@ def featurize_composition(df):
+        features.
+
+        Sets the `self.optimal_features` attribute to a list of feature
+        names.
 
         Args:
-            n:
-            full_cross_nmi: Allow to use a specific Cross-Features NMI.
+            n: number of desired features.
+            full_cross_nmi: specify the cross NMI between features as a
+                dataframe.
+
         """
         assert hasattr(self, 'df_featurized'), 'Please featurize the data first'
         assert not self.prediction, 'Please provide targets'
 
         ranked_lists = []
+        optimal_features_by_target = {}
 
         # Loading mutual information between features
         if full_cross_nmi is None:
             this_dir, this_filename = os.path.split(__file__)
             dp = os.path.join(this_dir, "data", "Features_cross")
-            full_cross_nmi = pd.read_pickle(dp)
+            if os.path.isfile(dp):
+                full_cross_nmi = pd.read_pickle(dp)
         else:
             full_cross_nmi = full_cross_nmi
 
-        for i,name in enumerate(self.names):
-            print("Starting target {}/{}: {} ...".format(i+1,len(self.targets),self.names[i]))
+        if full_cross_nmi is None:
+            if full_cross_nmi is None:
+                print('Computing cross NMI between all features...')
+                df = self.df_featurized.copy()
+                cross_nmi = get_cross_nmi(df)
+
+        for i, name in enumerate(self.names):
+            print("Starting target {}/{}: {} ...".format(i+1, len(self.targets), self.names[i]))
 
             # Computing mutual information with target
-            print("Computing mutual information ...")
+            print("Computing mutual information between features and target...")
             df = self.df_featurized.copy()
-            y_nmi = nmi_target(self.df_featurized,self.df_targets[[name]])[name]
+            y_nmi = nmi_target(self.df_featurized, self.df_targets[[name]])[name]
 
-            print('Computing optimal features...')
+            # remove columns from cross NMI if not present in feature NMI
             cross_nmi = full_cross_nmi.copy(deep=True)
+            missing = [x for x in cross_nmi.index if x not in y_nmi.index]
+            cross_nmi = cross_nmi.drop(missing, axis=0).drop(missing, axis=1)
 
-            a = []
-            for x in cross_nmi.index:
-                if x not in y_nmi.index:
-                    a.append(x)
-            cross_nmi = cross_nmi.drop(a,axis=0).drop(a,axis=1)
-            # opt_features = get_features_dyn(min(n,len(cross_nmi.index)),cross_nmi,y_nmi)
-            opt_features = get_features_dyn(min(n,len(y_nmi.index)),cross_nmi,y_nmi)
-            ranked_lists.append(opt_features)
-            print("Done with target {}/{}: {}.".format(i+1,len(self.targets),self.names[i]))
+            print('Computing optimal features...')
+            optimal_features_by_target[name] = get_features_dyn(min(n, len(y_nmi.index)), cross_nmi, y_nmi)
```

Comment:
Could now use get_features_relevance_redundancy here. Maybe some checks to be performed.

Reply:
Yep, my plan is to add tests for the current API then switch things over adiabatically :)

Reply:
Just to answer this more, the calculation of NMI needs to be reworked before we can test it easily, I think following the discussion yesterday we should try to merge this now so its not holding up other work.